tests/Kiota.Builder.Tests/Writers/CSharp/CodeMethodWriterTests.cs

```diff
@@ -0,0 +1,289 @@
+            };
+            parameter.Type = new CodeType(parameter) {
+                Name = "string"
+            };
+            method.MethodKind = CodeMethodKind.Serializer;
+            method.IsAsync = false;
+            AddSerializationProperties();
+            writer.Write(method);
+            var result = tw.ToString();
+            Assert.Contains("WriteStringValue", result);
+            Assert.Contains("WriteCollectionOfPrimitiveValues", result);
+            Assert.Contains("WriteCollectionOfObjectValues", result);
+            Assert.Contains("WriteEnumValue", result);
+            Assert.Contains("WriteAdditionalData(additionalData);", result);
+        }
+        [Fact]
+        public void WritesMethodAsyncDescription() {
+            
+            method.Description = methodDescription;
+            var parameter = new CodeParameter(method){
+                Description = paramDescription,
+                Name = paramName
+            };
+            parameter.Type = new CodeType(parameter) {
+                Name = "string"
+            };
+            method.AddParameter(parameter);
+            writer.Write(method);
+            var result = tw.ToString();
+            Assert.Contains("/// <summary>", result);
+            Assert.Contains(methodDescription, result);
+            Assert.Contains("<param name=", result);
+            Assert.Contains("</param>", result);
+            Assert.Contains(paramName, result);
+            Assert.Contains(paramDescription, result); 
+            Assert.Contains("</summary>", result);
+        }
+        [Fact]
+        public void WritesMethodSyncDescription() {
+            
+            method.Description = methodDescription;
+            method.IsAsync = false;
+            var parameter = new CodeParameter(method){
+                Description = paramDescription,
+                Name = paramName
+            };
+            parameter.Type = new CodeType(parameter) {
+                Name = "string"
+            };
+            method.AddParameter(parameter);
+            writer.Write(method);
+            var result = tw.ToString();
+            Assert.DoesNotContain("@returns a Promise of", result);
+        }
+        [Fact]
+        public void Defensive() {
+            var codeMethodWriter = new CodeMethodWriter(new CSharpConventionService());
+            Assert.Throws<ArgumentNullException>(() => codeMethodWriter.WriteCodeElement(null, writer));
+            Assert.Throws<ArgumentNullException>(() => codeMethodWriter.WriteCodeElement(method, null));
+        }
+        [Fact]
+        public void ThrowsIfParentIsNotClass() {
+            method.Parent = CodeNamespace.InitRootNamespace();
+            Assert.Throws<InvalidOperationException>(() => writer.Write(method));
+        }
+        [Fact]
+        public void ThrowsIfReturnTypeIsMissing() {
+            method.ReturnType = null;
+            Assert.Throws<InvalidOperationException>(() => writer.Write(method));
```

Comment:
If ReturnType is null and it is a requirement, we should throw ArgumentNullException

Reply:
I disagree, the return type is not an argument. So ArgumentNullException doesn't apply here. [InvalidOperationException](https://docs.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=net-5.0) "*The exception that is thrown when a method call is invalid for the object's current state.*" We're calling the method `Write` and passing it an object that's not valid for the method to succeed/do something that makes sense. I think this exception maps better.

Reply:
I now agree with you. I got overzealous and loose with ArgumentNullException.